File: solutions/typescript/rational-numbers/1/rational-numbers.ts
export class Rational {
  private a: number;
  private b: number;

  constructor(a: number, b: number) {
    this.a = a;
    this.b = b;
  }

  get numerator() {
    return this.a;
  }

  get denominator() {
    return this.b;
  }

  add(r2: Rational): Rational {
    const { a: a1, b: b1 } = this;
    const { a: a2, b: b2 } = r2;
    return new Rational(a1 * b2 + b1 * a2, b1 * b2).reduce();
  }

  sub(r2: Rational): Rational {
    const { a, b } = r2;
    return this.add(new Rational(a * -1, b));
  }

  mul(r2: Rational): Rational {
    const { a: a1, b: b1 } = this;
    const { a: a2, b: b2 } = r2;
    return new Rational(a1 * a2, b1 * b2).reduce();
  }

  div(r2: Rational): Rational {
    const { a: a1, b: b1 } = this;
    const { a: a2, b: b2 } = r2;
    return new Rational(a1 * b2, b1 * a2).reduce();
  }

  abs(): Rational {
    const { a, b } = this;
    return new Rational(Math.abs(a), Math.abs(b)).reduce();
  }

  exprational(n: number): Rational {
    let { a, b } = this;

    if (n < 0) {
      const temp = a;
      a = b;
      b = temp;
      n *= -1;
    }

    return new Rational(a ** n, b ** n).reduce();
  }

  expreal(x: number): number {
    const { a, b } = this;
    return (x ** a) ** (1 / b);
  }

  reduce(): Rational {
    const { a, b } = this;
    const gcd = this.gcd(a, b);

    let ra = a / gcd;
    let rb = b / gcd;

    if (rb < 0) {
      ra *= -1;
      rb *= -1;
    }

    return new Rational(ra, rb);
  }

  private gcd(a: number, b: number): number {
    if (a * b == 0) return Math.max(a, b);
    let gcd = Math.min(Math.abs(a), Math.abs(b));

    while (gcd > 1) {
      if (a % gcd == 0 && b % gcd == 0) break;
      gcd--;
    }

    return gcd;
  }
}

File: solutions/elixir/rational-numbers/1/lib/rational_numbers.ex
defmodule RationalNumbers do
  @type rational :: {integer, integer}

  @doc """
  Add two rational numbers
  """
  @spec add(a :: rational, b :: rational) :: rational
  def add(a, b) do
    {an, ad} = a
    {bn, bd} = b
    reduce({an * bd + ad * bn, ad * bd})
  end

  @doc """
  Subtract two rational numbers
  """
  @spec subtract(a :: rational, b :: rational) :: rational
  def subtract(a, b) do
    {an, ad} = a
    {bn, bd} = b
    reduce({an * bd - ad * bn, ad * bd})
  end

  @doc """
  Multiply two rational numbers
  """
  @spec multiply(a :: rational, b :: rational) :: rational
  def multiply(a, b) do
    {an, ad} = a
    {bn, bd} = b
    reduce({an * bn, ad * bd})
  end

  @doc """
  Divide two rational numbers
  """
  @spec divide_by(num :: rational, den :: rational) :: rational
  def divide_by(num, den) do
    {an, ad} = num
    {bn, bd} = den
    reduce({an * bd, ad * bn})
  end

  @doc """
  Absolute value of a rational number
  """
  @spec abs(a :: rational) :: rational
  def abs(a) do
    {num, den} = a
    reduce({Kernel.abs(num), Kernel.abs(den)})
  end

  @doc """
  Exponentiation of a rational number by an integer
  """
  @spec pow_rational(a :: rational, n :: integer) :: rational
  def pow_rational(a, n) when n < 0 do
    {num, den} = a
    pow_rational({den, num}, Kernel.abs(n))
  end

  def pow_rational(a, n) do
    {num, den} = a
    reduce({Integer.pow(num, n), Integer.pow(den, n)})
  end

  @doc """
  Exponentiation of a real number by a rational number
  """
  @spec pow_real(x :: integer, n :: rational) :: float
  def pow_real(x, n) do
    {a, b} = n
    :math.pow(:math.pow(x, a), 1 / b)
  end

  @doc """
  Reduce a rational number to its lowest terms
  """
  @spec reduce(a :: rational) :: rational
  def reduce(a) do
    {n, d} = a
    gcd = Integer.gcd(n, d)
    n = div(n, gcd)
    d = div(d, gcd)
    {if(d < 0, do: n * -1, else: n), Kernel.abs(d)}
  end
end
